File: typescript/src/useCase/OrderShipmentUseCase.ts
import OrderRepository from "../repository/OrderRepository";
import ShipmentService from "../service/ShipmentService";
import OrderShipmentRequest from "./OrderShipmentRequest";
import OrderCannotBeShippedError from "./OrderCannotBeShippedError";
import OrderCannotBeShippedTwiceError from "./OrderCannotBeShippedTwiceError";
import OrderStatus from "../domain/OrderStatus";

export default class OrderShipmentUseCase {
    private readonly _orderRepository: OrderRepository;
    private readonly _shipmentService: ShipmentService;

    constructor(orderRepository: OrderRepository, shipmentService: ShipmentService) {
        this._orderRepository = orderRepository;
        this._shipmentService = shipmentService;
    }

    public run(request: OrderShipmentRequest): void {
        const order = this._orderRepository.getById(request.orderId);

        if (order.status === OrderStatus.CREATED || order.status === OrderStatus.REJECTED) {
            throw new OrderCannotBeShippedError();
        }

        if (order.status == OrderStatus.SHIPPED) {
            throw new OrderCannotBeShippedTwiceError();
        }

        this._shipmentService.ship(order);

        order.status = OrderStatus.SHIPPED;
        this._orderRepository.save(order);
    }

}

File: csharp/OrderDispatchKata/UseCase/OrderShipmentUseCase.cs
﻿using OrderDispatchKata.Domain;
using OrderDispatchKata.Repository;
using OrderDispatchKata.Service;

namespace OrderDispatchKata.UseCase;

public class OrderShipmentUseCase
{
    private readonly OrderRepository orderRepository;
    private readonly ShipmentService shipmentService;

    public OrderShipmentUseCase(OrderRepository orderRepository, ShipmentService shipmentService)
    {
        this.orderRepository = orderRepository;
        this.shipmentService = shipmentService;
    }

    public void run(OrderShipmentRequest request)
    {
        var order = orderRepository.getById(request.getOrderId());

        if (order.getStatus().Equals(OrderStatus.CREATED) || order.getStatus().Equals(OrderStatus.REJECTED))
            throw new OrderCannotBeShippedException();

        if (order.getStatus().Equals(OrderStatus.SHIPPED)) throw new OrderCannotBeShippedTwiceException();

        shipmentService.ship(order);

        order.setStatus(OrderStatus.SHIPPED);
        orderRepository.save(order);
    }
}